src/pybind/mgr/volumes/fs_subvolume.py

```diff
@@ -0,0 +1,560 @@
+
+"""
+Copyright (C) 2019 Red Hat, Inc.
+
+LGPL2.1.  See file COPYING.
+"""
+
+import errno
+import json
+import logging
+import os
+
+from ceph_argparse import json_command
+
+import cephfs
+import rados
+
+def to_bytes(param):
+    '''
+    Helper method that returns byte representation of the given parameter.
+    '''
+    if isinstance(param, str):
+        return param.encode()
+    else:
+        return str(param).encode()
+
+class RadosError(Exception):
+    """
+    Something went wrong talking to Ceph with librados
+    """
+    pass
+
+
+RADOS_TIMEOUT = 10
+
+log = logging.getLogger(__name__)
+
+# Reserved subvolume group name which we use in paths for subvolumes
+# that are not assigned to a group (i.e. created with group=None)
+NO_GROUP_NAME = "_nogroup"
+
+
+class SubvolumePath(object):
+    """
+    Identify a subvolume's path as group->subvolume
+    The Subvolume ID is a unique identifier, but this is a much more
+    helpful thing to pass around.
+    """
+    def __init__(self, group_id, subvolume_id):
+        self.group_id = group_id
+        self.subvolume_id = subvolume_id
+        assert self.group_id != NO_GROUP_NAME
+        assert self.subvolume_id != "" and self.subvolume_id is not None
+
+    def __str__(self):
+        return "{0}/{1}".format(self.group_id, self.subvolume_id)
+
+
+class SubvolumeClient(object):
+    """
+    Combine libcephfs and librados interfaces to implement a
+    'Subvolume' concept implemented as a cephfs directory and
+    client capabilities which restrict mount access to this
+    directory.
+
+    Additionally, subvolumes may be in a 'Group'.  Conveniently,
+    subvolumes are a lot like manila shares, and groups are a lot
+    like manila consistency groups.
+
+    Refer to subvolumes with SubvolumePath, which specifies the
+    subvolume and group IDs (both strings).  The group ID may
+    be None.
+
+    In general, functions in this class are allowed raise rados.Error
+    or cephfs.Error exceptions in unexpected situations.
+    """
+
+    # Where shall we create our volumes?
+    DEFAULT_SUBVOL_PREFIX = "/subvolumes"
+    DEFAULT_NS_PREFIX = "fssubvolumens_"
+
+    def __init__(self, subvolume_prefix=None, pool_ns_prefix=None, rados=None,
+                 fs_name=None):
+        self.fs = None
+        self.fs_name = fs_name
+        self.connected = False
+
+        self.rados = rados
+
+        self.subvolume_prefix = subvolume_prefix if subvolume_prefix else self.DEFAULT_SUBVOL_PREFIX
+        self.pool_ns_prefix = pool_ns_prefix if pool_ns_prefix else self.DEFAULT_NS_PREFIX
+
+    def get_mds_map(self):
+        fs_map = self._rados_command("fs dump", {})
+        return fs_map['filesystems'][0]['mdsmap']
```

Comment:
This can be gotten `_volume_get_fs()`. Just need plumbing to give subvolume access to the mgr plugin.

Reply:
this method is actually not needed yet.